### ISSUE TITLE
Modelica stateSelection not supported

### ISSUE THREAD

**Issue by dospes** [pymoca/pymoca#255]
Hello!

tl;dr: The stateSelection features of Modelica are not supported by the pymoca parser. They are, in our opinion, not necessary for sympy (or casadi) models anyways, so they could be ignored when parsing.

We have been trying to use pymoca to convert flattened Modelica models (MOF-files) to sympy models. This works with very basic models. When flattening Models that are modeled using Blocks from the Modelica library, e.g. a mass (Modelica.Mechanics.Translational.Components.Mass), these flattened models include attributes and objects pertaining to stateSelection (see [https://www.claytex.com/blog/state-selection-overview-and-definitions/](https://www.claytex.com/blog/state-selection-overview-and-definitions/) for a quick overview what stateSelection is and why it is used). These attributes and objects are not supported by the parser, resulting in varying exceptions.
The stateSelection is, in our opinion however, not important for sympy models, as it is only used in Modelica when generating actual simulation binaries using a Modelica compiler (e.g. the OpenModelica Compiler omc). This being the case, our proposal is to ignore these attributes/objects when parsing the Modelica file.

Note, that we have been using #249, since our model also includes SI-units. The exceptions resulting from the units have been fixed by switching to this PR.

As we are not versed in parser programming, we do not know, how easy or difficult our suggestion would be to implement. This issue is therefore mainly for completeness sake, and so that our work on this matter could be further used to improve pymoca, should that be wished.

I will now describe the exceptions that we have encountered so far:

1.  The enumeration used for the stateSelect attribute ist not supported

In particular, line 15
`final parameter enumeration(never, avoid, default, prefer, always) mass1__stateSelect = StateSelect__default "Priority to use s and v as states";`
results in the parser warnings
`line 15:18 mismatched input 'enumeration' expecting IDENT`
`line 15:67 mismatched input ')' expecting ';'`
and leads to an incomplete ast tree, resulting in a later exception. The line causing the error is not important for the model and only used when generating a binary with omc, so it can be removed, fixing the parser warnings and the exception.

2. The attribute stateSelect is not supported by pymoca

Both lines 16 and 22
`Real mass1__s(quantity = "Length", unit = "m", start = -0.5, fixed = true, stateSelect = StateSelect.default) "Absolute position of center of component (s = flange_a.s + L/2 = flange_b.s - L/2)";`
and
` Real mass1__v(quantity = "Velocity", unit = "m/s", start = 0.0, fixed = true, stateSelect = StateSelect.default) "Absolute velocity of component";`
have an object with the stateSelect attribute. This results in the exception
`raise Exception("Trying to set unknown symbol property {}".format(argument.component.name))
Exception: Trying to set unknown symbol property stateSelect`
in the method modify_symbol(...), l. 750 in tree.py. The attribute stateSelect is not part of ast.Symbol.ATTRIBUTES. This attribute is also not important for the model, so we remove it from the flat model. This fixes the exception, but results in another problem not related to the stateSelection, so I will create another issue for it.

For completeness sake, here is the whole flat model before doing any of the edits described above:
[mass_spring_example.mof.txt](https://github.com/pymoca/pymoca/files/9931119/mass_spring_example.mof.txt)

**Comment by kentrutan** [comment]
Thanks for taking time to submit the issue! I'll take a look.

**Comment by kentrutan** [comment]
1. The enumeration you show above in the flattened model line 15 does not appear to be valid syntax according to the Modelica language spec, so I think Pymoca is correct in flagging a syntax error.
2. `stateSelect` is easy enough to add to the parser, but the SymPy backend has an issue with the model as you mentioned. It chokes on the `time < sine1__startTime` expression. It also will not handle the containing if expression.

The SymPy backend has not been actively developed and is very limited. I can try fix this particular issue, but if you are expecting it to handle more complicated models, you will likely run into additional issues. My priority is on improving Modelica language coverage in the parser first, so my response time on SymPy backend issues may be a bit slow, though I would like to see it developed.

**Comment by dospes** [comment]
Thank you for the quick response. I am unsure, how or why the non-valid Modelica code is generated, but as I said it is easy, to just remove the line. It would be nice to have the parser ignore lines that have non-valid Modelica syntax,

**Comment by dospes** [comment]
Did no mean to close the issue.

**Comment by dospes** [comment]
...

**Comment by dospes** [comment]
This should probably be a separate issue, but after removing the conditional statement from the offending line, it still does not behave properly, because the time variable is not declared in the model and cannot be accessed by the SymPy backend.

If I understood it properly, the time variable is a general Modelica variable that needs not be declared in a model, so it would be good to have this variable be part of the tree symbols all of the time or at least, when it is used in an equation. In my case, I could not find it in the symbols of the tree, but as part of one of the equations of the tree,

**Comment by kentrutan** [comment]
> It would be nice to have the parser ignore lines that have non-valid Modelica syntax, but I do not know, if this is generally useful or only useful for our application.

This is very, very difficult to implement in general. We definitely won't be doing this any time soon.

> It would help if logical expressions were supported, but we have a simpler working example which works for our application, so it is not super important for us. If this turns out to be something we want to use more frequently, we might try to improve the SymPy backend ourselves, but for now it is only a proof of concept.

The SymPy backend is itself is a proof of concept and very limited. Take a look at [`test/gen_sympy_test.py`](https://github.com/pymoca/pymoca/blob/169b68506a1d0cd3e03af1dbd42a90155c6d3afb/test/gen_sympy_test.py) for examples of what it can do. It would be great to have someone contribute to improving it.

**Comment by kentrutan** [comment]
> This should probably be a separate issue, but after removing the conditional statement from the offending line, it still does not behave properly, because the time variable is not declared in the model and cannot be accessed by the SymPy backend.
> 
> If I understood it properly, the time variable is a general Modelica variable that needs not be declared in a model, so it would be good to have this variable be part of the tree symbols all of the time or at least, when it is used in an equation. In my case, I could not find it in the symbols of the tree, but as part of one of the equations of the tree,

As you have found, `time` is not implemented in the SymPy backend. Please do create a separate issue for it.

For your proof of concept, I recommend that you change the input to work within the current SymPy backend capabilities. If there is something critical that you cannot work around, then create an issue and note the criticality in the issue. If is something I can fix quickly, I might give it a shot, but I do not have much time to work on SymPy right now.
